src/main/java/com/salesforce/dataloader/security/EncryptionAesUtil.java

```diff
@@ -0,0 +1,255 @@
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import java.util.Locale;
+
+import javax.crypto.Cipher;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * com.salesforce.dataloader.security
+ *
+ * @author xbian
+ */
+public class EncryptionAesUtil {
+
+    private static final Logger LOGGER = Logger.getLogger(EncryptionAesUtil.class);
+
+    public enum OSType {
+        Windows, MacOS, Linux, Other
+    }
+
+    ;
+    protected static OSType detectedOS;
+
+    /**
+     * detect the operating system from the os.name System property and cache the result
+     *
+     * @returns - the operating system detected
+     */
+    public static OSType getOperatingSystemType() {
+        if (detectedOS == null) {
+            String OS = System.getProperty("os.name", "generic").toLowerCase(Locale.ENGLISH);
+            if ((OS.indexOf("mac") >= 0) || (OS.indexOf("darwin") >= 0)) {
+                detectedOS = OSType.MacOS;
+            } else if (OS.indexOf("win") >= 0) {
+                detectedOS = OSType.Windows;
+            } else if (OS.indexOf("nux") >= 0) {
+                detectedOS = OSType.Linux;
+            } else {
+                detectedOS = OSType.Other;
+            }
+        }
+        return detectedOS;
+    }
+
+    // Support single text encryption and decryption
+
+    static public Cipher cipher;
+    static public byte[] cipherKey;
+
+    // 16 bytes was used for 128 bit AES encryption
+    public static final int ENCRYPTION_KEY_LENGTH_IN_BYTES = 16;
+    public static final String DEFAULT_KEYFILE_NAME = "dataLoader.key";
+
+    static {
+        try {
+            cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
+            detectedOS = getOperatingSystemType();
+
+        } catch (Exception e) {
+            LOGGER.error("Fail to initialize encryption: " + e.getMessage());
+            throw new RuntimeException("Fail to initialize encryption:  ", e);
+        }
+    }
+
+
+    private byte[] extractIvBytes(byte[] cipheredText) {
+        byte[] iv = new byte[ENCRYPTION_KEY_LENGTH_IN_BYTES];
+        System.arraycopy(cipheredText, 0, iv, 0, ENCRYPTION_KEY_LENGTH_IN_BYTES);
+        return iv;
+    }
+
+    private byte[] extractCipherContent(byte[] cipheredText) {
+        byte[] cipheredPassword = new byte[cipheredText.length - ENCRYPTION_KEY_LENGTH_IN_BYTES];
+        System.arraycopy(cipheredText, ENCRYPTION_KEY_LENGTH_IN_BYTES, cipheredPassword, 0, cipheredText.length - ENCRYPTION_KEY_LENGTH_IN_BYTES);
+        return cipheredPassword;
+
+    }
+
+    private byte[] concatenateByteArray(byte[] a, byte[] b) {
+
+        byte[] c = new byte[a.length + b.length];
+        System.arraycopy(a, 0, c, 0, a.length);
+        System.arraycopy(b, 0, c, a.length, b.length);
+        return c;
+    }
+
+    public byte[] generateEncryptionKey() {
```

Comment:
I would expect this function to always set `cipherKey`, rather than returning an array that should (but might not always) be assigned to `cipherKey`. There's not any reason I can see to ever have key data in any array other than `cipherKey` and it runs the risk of data being encrypted/decrypted with the wrong key, or the wrong key bytes being written to a keyfile.

A separate function that generates an IV could be used for IV; using this function for an IV is arguably incorrect from a code maintainability perspective anyhow (keys are not IVs, even though they are often generated the same way).